implement Array#shuffle in ruby land

KBuild opened the pull request:
Hello!

Thanks to create this project!

I implement `Array#shuffle` and `Array#shuffle!`

I also know that its going to Rust implementation

But I think this logic can be made in Ruby and it's simple

If you think this PR makes sense, please accept it

If you don't, please close this PR

I really have expecting in this project!

Thanks

KBuild committed:
implement Array#shuffle in ruby land

lopopolo commented:
Hi @KBuild thanks for the PR. I'm acking that I've seen it, but I haven't had time to give it a review yet.

I noticed the Ruby linter is failing. You should be able to fix this by running `bundle exec rake` on your local branch.

KBuild committed:
fix lint

KBuild commented:
Thank you for reaction to my PR!

I fix that lint error!

@lopopolo

KBuild committed:
run ci

lopopolo commented on artichoke-backend/src/extn/core/array/array.rb:
```diff
@@ -1105,12 +1105,20 @@ def select!(&block)
 
-  def shuffle(_rng = (not_set = true)) # rubocop:disable Lint/UselessAssignment
-    raise NotImplementedError, 'TODO implement in Rust'
+  def shuffle(rng = nil)
```
the default assignment initializer is required because passing `nil` to `Array#shuffle` should raise an exception.

When `not_set` is `true`, `Array#shuffle` should use `Random::Default`

lopopolo commented on artichoke-backend/src/extn/core/array/array.rb:
```diff
@@ -1105,12 +1105,20 @@ def select!(&block)
-    raise NotImplementedError, 'TODO implement in Rust'
+  def shuffle(rng = nil)
+    rng_instance = rng.nil? ? Random.new : rng
+    shuffled_orders = (0..size - 1).map.with_index { |_n, i| [rng_instance.rand, i] }.sort { |a, b| a.first <=> b.first }
```
this is a clever approach.

lopopolo commented on artichoke-backend/src/extn/core/array/array.rb:
```diff
@@ -1105,12 +1105,20 @@ def select!(&block)
 
-  def shuffle(_rng = (not_set = true)) # rubocop:disable Lint/UselessAssignment
-    raise NotImplementedError, 'TODO implement in Rust'
+  def shuffle(rng = nil)
+    rng_instance = rng.nil? ? Random.new : rng
+    shuffled_orders = (0..size - 1).map.with_index { |_n, i| [rng_instance.rand, i] }.sort { |a, b| a.first <=> b.first }
```
This suggestion adds support for `[].shuffle` using the default `Random` instance and makes a few performance optimizations. Let me know what you think @KBuild.

```suggestion
  def shuffle(rng = (not_set = true))
    rng = Random::DEFAULT if not_set
    shuffled_orders = (0...size).map { |idx| [rng.rand, idx] }.sort { |a, b| a[0] <=> b[0] }
```

KBuild replied:
Thank you for informing suggestion!
I prefer the code keeping consistency rules
I will fix that

KBuild replied:
Oh, I check some document and find `Array#shuffle` in this link
https://docs.ruby-lang.org/en/2.6.0/Array.html#method-i-shuffle

It seems like to using `random` keyword arguments
So I think we should use

```rb
def shuffle(random: (not_set = true))
...
end
def shuffle(random: (not_set = true))
...
end
```

How about this?

I'm going to commit this idea, and please check code 🙏

lopopolo replied:
Looks good. Thanks for checking the docs.

lopopolo reviewed (changes requested):
Thanks @KBuild. Sorry this took a while for me to review. This looks great and I'm glad you found a way to do this in pure Ruby. I have one change I'd like to make: MRI uses the `Random::DEFAULT` RNG when a parameter is not given to `Array#shuffle`.

lopopolo commented:
It looks like the failing Documentation build is due to a nightly regression that was fixed upstream. Once you push a new commit with the `Random::DEFAULT` fix, the build should fix itself.

KBuild committed:
fix argument controlling

- use not_set keyword for check argument existance
- use keyword argument

lopopolo commented on artichoke-backend/src/extn/core/array/array.rb:
```diff
@@ -1105,12 +1105,25 @@ def select!(&block)
-  def shuffle(_rng = (not_set = true)) # rubocop:disable Lint/UselessAssignment
-    raise NotImplementedError, 'TODO implement in Rust'
+  def shuffle(random: (not_set = true))
+    random = Random::DEFAULT if not_set
```
```suggestion
    random = Random::DEFAULT if not_set # rubocop:disable Lint/DeprecatedConstants
```

lopopolo commented on artichoke-backend/src/extn/core/array/array.rb:
```diff
@@ -1105,12 +1105,25 @@ def select!(&block)
 
-  def shuffle(_rng = (not_set = true)) # rubocop:disable Lint/UselessAssignment
-    raise NotImplementedError, 'TODO implement in Rust'
+  def shuffle(random: (not_set = true))
```
thanks for catching this :)

lopopolo commented on artichoke-backend/src/extn/core/array/array.rb:
```diff
@@ -1105,12 +1105,25 @@ def select!(&block)
+    return self if length <= 1
+
+    self[0, length] =
+      if not_set
+        shuffle
+      else
+        shuffle(random: random)
+      end
```
looks good here thank you.

lopopolo committed:
Suppress deprecated constant lint for Random::DEFAULT

`Random::DEFAULT` was depreated in Ruby 3.0.0, but
we're implementing Ruby 2.6.3.

lopopolo commented:
Thanks for the updates @KBuild. I've added one commit on top of your branch that addresses a RuboCop deprecation warning. Assigning to myself to make sure this gets merged.

Thanks for contributing to Artichoke! 💎
